File: Graph/*****Cycle Detection in Directed Graph using BFS(Kahn's Algorithm).cpp
Approach:

1. If we can get topologic sort then we can say graph is DAG, else graph is cyclic.

#CODE:
bool isCyclic(int n, vector<int> adj[]) {
    vector<int> indegree(n,0);
    for(int i=0;i<n;++i) {
        for(auto it: adj[i]) {
            indegree[it]++;
        }
    }
    queue<int> q;
    int count = 0;
    for(int i=0;i<n;++i) {
        if(indegree[i] == 0) {
            q.push(i);  
        }
    }
    while(!q.empty()) {
        int node = q.front();
        q.pop();
        count++;
        for(auto it: adj[node]) {
            indegree[it]--;
            if(indegree[it] == 0) {
                q.push(it);
            }
        }
    }
    return count != n;
}
=> Time Complexity = O(N + E)
=> Space Complexity = O(N)

File: Graph/*****Cycle Detection in Directed Graph using BFS(Kahn's Algorithm).java
Approach: 
1. If we can get a topological sort of the graph then graph is DAG(Directed Acyclic Garph)

#CODE:

class Solution {
    public boolean isCyclic(int V, ArrayList<ArrayList<Integer>> adj) {
        int indegree[] = new int[V];
         int ans[] = new int[V];
         Arrays.fill(indegree, 0);
         for(int i=0;i<V;++i) {
             for(Integer it: adj.get(i)) {
                 indegree[it]++;
             }
         }
         
         Queue<Integer> q = new LinkedList<>();
         for(int i=0;i<V;++i) {
             if(indegree[i] == 0) {
                 q.offer(i);
             }
         }
         
         int cnt = 0;
         int idx = 0;
         while(!q.isEmpty()) {
             int node = q.poll();
             ans[idx++] = node;
             cnt++;
             for(Integer it: adj.get(node)) {
                 indegree[it]--;
                 if(indegree[it] == 0) {
                     q.offer(it);
                 }
             }
         }
         return cnt == V ? false : true;
    }
}